feat: Virtual DocFields

gavindsouza opened the pull request:
### Virtual DocFields

Consider adding a property to your DocType controller, like age to Person as shown below

```python
class Person(Document):
    @property
    def age(self):
        return frappe.utils.now_datetime() - self.creation
```

Before this, _age_ would show up as `None`. But, through this, the value is set nicely when accessed via `as_dict` / REST. 

If you have a `@property` in the controller class, this data used to not be accessible anywhere other than directly through the object via `frappe.get_doc`.

<img width="1440" alt="Screenshot 2021-10-27 at 12 05 25 PM" src="https://user-images.githubusercontent.com/36654812/139035135-8157c96f-b00f-4716-82aa-e0c3b7932415.png">

I added a read-only Data **virtual** field to Note, called it difference and added the following block in the Note controller.

```python
@property
def difference(self):
	return self.modified - self.creation
```


<img width="1440" alt="Screenshot 2021-10-27 at 12 05 08 PM" src="https://user-images.githubusercontent.com/36654812/139035108-5e3732e1-51de-452e-8c45-235c6d6e3a3c.png">

Made changes to raise via `check_fieldname_conflicts` to ignore virtual fields. Although, a msgprint will be shown as a nudge/reminder.

ref: https://github.com/frappe/frappe/issues/14462#issuecomment-952576498

> This PR introduces a Virtual check for DocFields, similar to DocTypes.

Docs: https://frappeframework.com/docs/v13/user/en/basics/virtual_docfield

gavindsouza committed:
refactor!: Remove dead methods from PropertySetter controller

gavindsouza committed:
fix: Handle Virtual DocFields in Desk

* Show as "Read Only" in forms/controls/webforms etc
* Exclude from filtering options in views
* Exclude from quick entry

gavindsouza committed:
fix: Calculate dynamic properties for as_dict

If you have a @property in the controller class, this data used to not
be accessible anywhere other than directly through the object via
`frappe.get_doc`. This fix changes that.

gavindsouza committed:
refactor: check_fieldname_conflicts

- Don't raise if docfield is read only. This allows for dynamic data
fields whose values don't need to be stored in the database
- After 1599f422d320122391acacf91a2206ba42165517, it's possible to pass
  dynamic properties' data to Desk/APIs

This makes it possible to show Person.age on Desk/Api if you have their
DOB saved in the table

daraul commented:
I hope this is going to get done. Do you need help?

surajshetty3416 commented:
@gavindsouza  Please re-open once this is ready for review.

gavindsouza commented:
@daraul I'll try to get this done this week. If not, I don't think I'll be able to look into this, this month. 😅

daraul commented:
I get it, dude. I'm swamped right now, too.

gavindsouza commented:
Closing since I couldn't take this further this week. Will attempt to clear this one in a couple of weeks.

gavindsouza committed:
fix: Don't add property value if NoneType or Falsy

gavindsouza commented:
Rebased, refactored backup tests while I was at it. Realized that should be in another PR so moved it out from here.

gavindsouza committed:
test: Add test for dynamic docfield

gavindsouza commented:
For the sake of this test, I have added a read-only Data field but instead of adding a varchar column that's not going to be used, I think we should have the concept of virtual docfields too 🤔.

daraul commented:
A virtual docfield being one that ties to a function, or combination of other docfields, instead of a column? I agree, that sounds like the right way to go.

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into dynamic-docfields

gavindsouza committed:
feat: Virtual DocField

Add virtual docfield through which you can make callable properties for
the controller

gavindsouza committed:
fix: Add is_virtual to Custom Field

The diff for this commit is too large because the format was changed.
But essentially, only "Is Virtual" field was added. This commit was
meant to be a part of the preceeding one. It was kept separate only
because the diffs too large :')

gavindsouza commented:
This is how it looks now, if "Is Virtual" is checked for the field and a controller property exists, you may access it via APIs. I added a field called "difference" to Note (through Customize Form) and added a corresponding property in the controller such as:

```python
@property
def difference(self):
	return self.modified - self.creation
```

And this is how it worked:

<img width="512" alt="Screenshot 2022-01-27 at 10 01 49 PM" src="https://user-images.githubusercontent.com/36654812/151409212-b63a5dab-9cab-4fc3-ac3d-5470320671bc.png">

gavindsouza committed:
fix(test): Update test_virtual_fields

gavindsouza committed:
fix: Skip raising only if virtual docfield while checking for conflicts

surajshetty3416 reviewed (changes requested):
@gavindsouza tried this out... works fine. Few points.

1. Virtual field should ideally not create a column in the database since we are not saving any data.
2. Also, let's use "options" of the virtual field as "expression" input so that custom doctypes with no controller file can use this as well?
**Example:**
	**Configuration:**
	<img width="1030" alt="Screenshot 2022-01-31 at 9 45 51 AM" src="https://user-images.githubusercontent.com/13928957/151738579-a976c86b-c11b-40d2-aaa3-a6cd9b81bb21.png">
	**Output:**
	<img width="549" alt="Screenshot 2022-01-31 at 9 44 21 AM" src="https://user-images.githubusercontent.com/13928957/151738193-2aa3fed2-e531-4cab-94b3-9bd5a92a9e9e.png">
	**Code change required to make it work:**
	<img width="847" alt="Screenshot 2022-01-31 at 9 39 45 AM" src="https://user-images.githubusercontent.com/13928957/151738400-940feef9-13ad-4613-9b80-d99cf371dd90.png">
3. Shouldn't fields only with "Read Only" fieldtype should get the "Is Virtual" check option?

gavindsouza committed:
fix: Don't create DB column for Virtual DocFields

gavindsouza committed:
fix: Don't validate fieldtype changes for Virtual DocFields

gavindsouza committed:
feat(minor): Allow expressions in options for DocFields

This means you don't have to write a custom controller or change backend
code to use Virtual DocFields! Write it in the options column for the
virtual field in Customize Form

Added doc and safe globals (from safe_exec) into the evaluation
namespace

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into dynamic-docfields

gavindsouza commented:
Keeping in mind the suggestions by @surajshetty3416, I expanded it slightly to allow writing code directly in options

<img width="1239" alt="Screenshot 2022-02-01 at 6 06 55 PM" src="https://user-images.githubusercontent.com/36654812/151970887-bcb5190c-b1e7-4958-83cf-e86ee44f7828.png">

<img width="1239" alt="Screenshot 2022-02-01 at 6 07 18 PM" src="https://user-images.githubusercontent.com/36654812/151970900-9e8cdf4e-5edd-4205-b13e-f368d61c7745.png">

gavindsouza committed:
fix: Skip data field validations for virtual field

gavindsouza committed:
test: Add test for docfield.options virtual df usage

surajshetty3416 commented:
Looks ready to merge... let's also add docs for this since Customize Form can be used to configure this feature.

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into dynamic-docfields

surajshetty3416 commented:
@gavindsouza Merging this, hoping you are already working on the docs.

gavindsouza commented:
@surajshetty3416 Docs added on https://frappeframework.com/docs/v14/user/en/basics/virtual_docfield

jalajc commented:
@gavindsouza 
Thanks for bringing a new concept to frappe. I found this very common in other ERP/Enterprise applications.
However, I have few observations:

I think these should be rather named as `Calculated Fields`. They are calculated by various mechanisms.

For virtual fields, per name say, scope should be much more open as they should be available for UI manipulation (and not read only).
However, it should be possible for any other field, that is dependent on virtual field or their combination, to be stored in a datable table->column.
Consider a case:
 - Have couple of setting fields (which exist on UI) 1. Checkbox, 2. Select field, 3. data field. 
 - This setting fields on UI behave as natural fields , say , you can check and uncheck, select dropdown, set a string in data field.
 - However, I don't want to store them in database.
 - Rather, I want to generate a string combination of this setting fields based on their UI state and store the generated combination string in a (hidden) field mapped to database->table->column (kind of json having key-pair store).

So essentially I am talking about two paradigms here...
1. A purely UI based field  ( I personally prefer calling them virtual Unlike in this PR) -- no database column for them but have option of readonly or editable.
2. A Calculated field (which is referred as virtual field in this PR) with possibility of mapping a column or no-column in database. Read-only (but could be hidden). 

Tagging @rmehta for his view and comments.

rmmr commented:
I would like to bring up the current state of the implementation as imo it's rather lacking. First issue that I immediately encountered is the List View breaking when a Virtual field is displayed there.

I think including functionality like this is great, but it needs to be done carefully. Currently in the BaseList implementation all fields are immediately converted into a column names (which imho is not something that should be done on the client side). Secondly db_query.py is pretty much directly invoked, so obviously Virtual Fields will fail here. 

I believe that a rather comprehensive rework of the way this is done is needed in order to even mention support for Virtual Fields in the docs. Or state that it only works in very restrictive scenario's. 

I'm willing to look into a rework of this. Also further points of attention:

- Virtual fields can't be used to sort in list views
- Virtual fields can't be used to filter in list views
- How would virtual fields behave with fieldtype of Link?

A few days ago i worked on  pull request #17130, which is not directly related but it has shown some of the limitations currently in Frappe, especially in the way list views work. I believe that a new layer should be added between client views and db_query.py which is responsible for identifying the type of field and it's dependencies (in case of link titles as shown in #17130) and perform the correct query.

In case of the Virtual field this would work as followed:

- Client side code no longer converts fields name into database column names!
- frappe.desk.reportview.get won't directly call db_query.py but instead call this new layer
- This new layer will inspect the client "query" and uses DocType and DocField meta to write the correct query for db_query.py
- It will omit virtual fields from the columns
- It will find the related columns for link_titles as #17130 does on client side now
- Now since virtual fields **can depend on column values that are not inside the list view query**, we introduce either a decorator on the Doctype controller method for that Virtual field, or we utilize the DocField's options to specify **dependant columns for this field.** To go even further we could allow virtual fields to depend on each-other.
- we execute the db query. And in the new layer we call the doctype controllers methods for the virtual fields.
- return this result to the client.

magic-overflow commented:
Virtual Docfield shouldn't be tracked changes.
